stdlib/domain.ml

```diff
@@ -151,31 +163,35 @@ let do_at_first_spawn () =
     first_spawn_function := (fun () -> ())
   end
 
-let exit_function = Atomic.make (fun () -> ())
+let at_exit_key = DLS.new_key (fun () -> (fun () -> ()))
 
-let rec at_exit f =
-  let wrapped_f () = try f () with _ -> () in
-  let old_exit = Atomic.get exit_function in
-  let new_exit () = wrapped_f (); old_exit () in
-  let success = Atomic.compare_and_set exit_function old_exit new_exit in
-  if success then
-    Stdlib.at_exit wrapped_f
-  else at_exit f
+let at_exit f =
+  if is_main_domain () then
+    Stdlib.at_exit f
+  else begin
+    let old_exit : unit -> unit = DLS.get at_exit_key in
+    let new_exit () = f (); old_exit () in
+    DLS.set at_exit_key new_exit
+  end
```

Comment:
I wasn't expecting this special case for the main domain.  If a domain other than the main domain calls `Stdlib.exit`, it's going to run exit handlers not associated with itself...  

An alternative could be for `Stdlib.exit` to run the at-exit handlers for the calling domain (those registered with `Domain.at_exit`), then the process-wide handlers registered with `Stdlib.at_exit`.

Reply:
I wasn't paying attention to `Stdlib.exit`. With https://github.com/ocaml/ocaml/pull/11213/commits/86627bf875e387194f704d94ce0829e9b627f258, `Stdlib.exit` calls domain-local callbacks and then the process-wide callbacks registered with `Stdlib.at_exit`. This also removes the special casing for main domain in `Domain.at_exit`.

I discovered a separate issue with exceptions raised by `at_exit` callbacks and `Stdlib.exit`. I've made an issue for this: https://github.com/ocaml/ocaml/issues/11221.

Reply:
Let me know what you think about the latest commit https://github.com/ocaml/ocaml/commit/86627bf875e387194f704d94ce0829e9b627f258.